### ISSUE TITLE
Suggestion: "Better" Junction Table Support

### ISSUE THREAD

**Issue by MethyleneGaming** [yiisoft/yii2#2900]
![example](https://cloud.githubusercontent.com/assets/549301/2544859/c1f28d3e-b607-11e3-90e6-58e5f0ba2db1.jpg)
Users Model:

``` php
class Users extends \yii\db\ActiveRecord
{
    public function getGroups()
    {
        return $this->hasMany(Groups::className(), ['id' => 'group_id'])
            ->viaTable('tbl_users_groups', ['user_id' => 'id'])
            ->pivot('unique_1', 'unique_2', 'created', 'updated');
    }
}
```

Users Controller:

``` php
class TestController extends \yii\web\Controller
{
    public function actionGroup()
    {
        $users = Users::find()
            ->with('groups')
            ->all();

        foreach ($users as $user) {
            foreach ($user->groups as $group) {
                if ($group->pivot->unique_1 === "testValue") {
                    echo $group->pivot->unique_1;
                    echo $group->pivot->unique_2;
                    echo $group->pivot->created;
                    echo $group->pivot->updated;                
                }
            }
        }
    }
}
```

http://laravel.com/docs/eloquent#working-with-pivot-tables

This has been taken from the Laravel framework, I was wondering if we could get this type of functionality in the Yii2 Base Framework, and if the developers do not want it to work this way for whatever reason, could someone tell me what is the best way to achieve this type of functionality with what the yii2 framework currently has? The most efficient way that is. Currently I would say this::

Users Model:

``` php
class Users extends \yii\db\ActiveRecord
{
    public function getUsersGroups()
    {
        return $this->hasMany(UsersGroups::className(), ['user_id' => 'id']);   
    }

    public function getGroups()
    {
        return $this->hasMany(Groups::className(), ['id' => 'group_id'])
            ->via('usersGroups');
    }
}
```

UsersGroups Model:

``` php
class UsersGroups extends \yii\db\ActiveRecord
{
    public function getGroup()
    {
        return $this->hasOne(Groups::className(), ['id' => 'group_id']);
    }
}
```

Users Controller:

``` php
class TestController extends \yii\web\Controller
{
    public function actionGroup()
    {
        $users = Users::find()
            ->with('usersgroups.group')
            ->all();

        foreach ($users as $user) {
            foreach ($user->usersgroups as $userGroup) {
                if ($userGroup->unique_1 === "testValue") {
                    echo $userGroup->group->unique_1;
                    echo $userGroup->group->unique_2;
                    echo $userGroup->group->created;
                    echo $userGroup->group->updated;                
                }


            }
        }
    }
}
```

I ask this because I noticed that there was an issue that "fixed" pivot table's models not being generated by default so it seems as though the second method, currently what I use is not really the intended method to be used in this situation.

**Comment by saidbakr** [comment]
Many to Many relations regarding saving and updating, specially, using the ActiveForm widget is a nightmare. I'm not sure, if there is missing or buggy features, or it is limitation in the documentation.

For instance, suppose the following view:
User update or create form with multiple select options list of user's groups to attach or deattach groups to the user via the conjunction table `pivot`!

**Comment by samdark** [comment]
@saidbakr so what do you suggest?

**Comment by johonunu** [comment]
@saidbakr There is an extension for your issue:
https://github.com/la-haute-societe/yii2-save-relations-behavior

**Comment by saidbakr** [comment]
@samdark What I am looking for is a simple implementation that make the process more easy or more clear. I suggest trying to make something [**like that**](http://laraveldaily.com/pivot-tables-and-many-to-many-relationships/) in Yii2.

**Comment by samdark** [comment]
@saidbakr

**Comment by samdark** [comment]
That worth thinking about. Are you concerned about defining relations, using them or about updates?

**Comment by saidbakr** [comment]
@samdark About update, indeed, the mass of the problem lies in the update, secondly, create.

**Comment by samdark** [comment]
So I did a little research.

## Add another product to the current shop instance

```php
// Laravel
$shop = Shop::find($shop_id);
$shop->products()->attach($product_id);

// Yii
$product = Product::findOne($product_id);
$shop = Shop::findOne($shop_id);
$shop->link('products', $product);
```

Both are OK.

## Remove a product from the shop

```php
// Laravel
$shop->products()->detach($product_id);

// Yii
$product = Product::findOne($product_id);
$shop->unlink('products', $product);
```

Both are OK.

## Remove all products from a particular shop:

```php
// Laravel
$shop->products()->detach();

// Yii
$shop->unlinkAll('products');
```

Both are OK.


## Add or remove products with particular ID

```php
// Laravel
$shop->products()->attach([123, 456, 789]);
$shop->products()->detach([321, 654, 987]);

// Yii
$products = Product::findAll([123, 456, 789]);
foreach ($prodcuts as $product) {
    $shop->link('products', $product);
}

$products = Product::findAll([321, 654, 987]);
foreach ($prodcuts as $product) {
    $shop->unlink('products', $product);
}
```

Yii requires a bit more effort. There's extra syntax in Laravel `$shop->products()->attach([123, 456, 789 => ['price' => 49.99]]); ` for adding extra data to particular items.

## Replace products with a set of IDs

```php
// Laravel
$product->shops()->sync([1, 2, 3]);

// Yii
$shop->unlinkAll('products');
$products = Product::findAll([1, 2, 3]);

foreach ($prodcuts as $product) {
    $shop->link('products', $product);
}
```

A shortcut could be added to Yii.

## Replace products with a set of IDs without detaching existing ones

```php
// Laravel
$product->shops()->syncWithoutDetaching([1, 2, 3]);

// Yii, method 2
$products = Product::findAll([1, 2, 3]);
// find what actually was added/remove and avoid touching existing rows
// way lengthy
```

A shortcut could be added to Yii. Seems there are extensions for it already:

https://github.com/voskobovich/yii2-linker-behavior#support

### Add non-existing, remove existing

```php
// Laravel
$product->shops()->toggle([1, 2, 3]);

// Yii
$ids = [1, 2, 3];
$shopProducts = ShopProduct::find()->where(['product_id' => $ids])->indexBy('product_id');

foreach ($ids as $id) {
    if (isset($shopProducts[$id])) {
        $shop->unlink('products', $shopProducts[$id] );
    } else {
        $shop->link('products', Product::findOne($id));
    }    
}
```

Not sure what's use case for this method. Looks weird.

## Attach with setting additional params to junction table

```php
// Laravel
$shop->products()->attach(1, ['products_amount' => 100, 'price' => 49.99]);

// Yii
$product = Product::findOne(1);
$shop->link('products', $product, ['products_amount' => 100, 'price' => 49.99]);
```

Both are OK.

**Comment by samdark** [comment]
Thoughts?

**Comment by devmustafa** [comment]
I suggest `link() & unlink()` methods to work with findAll(), so for example:
```
$products = Product::findAll([123, 456, 789]);
// adding
$products->link('products', $product, $params = [], $sync = false);
// removing
$products->unlink('products', $product, $params = [], $sync = false);
```
Will be more neat.

**Comment by samdark** [comment]
I'd rather not mix these and introduce separate methods.

**Comment by saidbakr** [comment]
@samdark About **Replace products with a set of IDs without detaching existing ones**, I'd like to confirm that it should be basic in the framework not by an extension.

Secondly, after testing and releasing those features we should work on, clearly, demonstrating them in the documentation.
Thank you for your efforts.

**Comment by samdark** [comment]
> I'd like to confirm that it should be basic in the framework not by an extension

Agree. I just referenced extension because both description and implementation there looks good.

**Comment by samdark** [comment]
@saidbakr do you have ideas for method signatures?

**Comment by saidbakr** [comment]
@samdark I suggest to be `pivot` as follows:

```
pivot(Array relatedModel, [string pivotTableName])

```
The `relatedModel` is an associative array with one key `string model name` and its value is an array of data should be added/synchronized to the pivot table for example, it,  `relatedModel`, may take two forms as follows :

```
$relatedModel = [
   'Shop' => [2,3,4]
]
```
For the above form, suppose we work on a Product model of `id` equals `10` so this data should add\sync that product to shops 2,3,4 that's make the pivot table have the following data after save:
```
shop_id | product_id
2   => 10
3  => 10
4  => 10
```
**the second form of the array** should maintain  the extra data in the pivot like `amount` as follows:

```
$relatedModel = [
  'Shop' => [
     '2' => ['amount' => 120,...],
     '3' => ['amount' => 19,...],
     '4' =>  ['amount' => 23,...],
  ]
]
```
**pivotTableName** optional if the table name is not obey conventions. In later phase we could introduce third optional boolean parameter that control weather delete absent attribute or not i.e (sync or attach) in other words, by default it should perform syncing.

**Comment by samdark** [comment]
@saidbakr "pivot" is a wrong term misused by Laravel and Yii 1.1. Correct term is "junction".

**Comment by saidbakr** [comment]
@samdark Although I don't know why pivot is wrong term, however, let it to be `junction` or `relation` Indeed to works on the relation mainly.

**Comment by samdark** [comment]
https://en.wikipedia.org/wiki/Pivot_table

**Comment by saidbakr** [comment]
Indeed, Junction table, *and as I've understood from the wiki link that you have supplied*,  could be regarded as a mean of filtration, exactly like pivot do. In other words, pivot table or the junction table does not introduce new entity.

However, when we use it to handle extra data, and at this point, it may be regarded as a new entity, so the filtration principle, hardly, applied there. i.e pivot is a special case of junction. I hope that I could understand this well.

**Comment by cebe** [comment]
linking https://github.com/yiisoft/yii2-collection/issues/8 as this could probably be solved in the collection extension.

**Comment by saidbakr** [comment]
@cebe Could you please show how could it be solved? I looked at the Change Log of 2.0.13 and I could not find it.

**Comment by samdark** [comment]
Won't be implemented for 2.0.
